binary-data-formats.qmd

```diff
@@ -0,0 +1,404 @@
+There are no hard and fast rules about how to partition a data set: the results will depend on your data, access patterns, and the systems that read the data.
+As a rough guide, the arrow package documentation suggests the following heuristics:
+
+-   Avoid files smaller than 20MB and larger than 2GB.
+-   Avoid partitioning layouts that produce more than 10,000 distinct files.
+
+According to these guidelines, the Seattle library data is a little too large to be stored as a single 4GB parquet file.
+We should consider splitting it up.
+
+As a general rule, a useful strategy for partitioning a large dataset is to create meaningful subsets based on a variable that you're likely to query on.
+For the library data it seems likely that we would often want to analyze the data separately by year, so the `CheckoutYear` column seems like a good candidate.
+
+Partitioning a data set can be done using the `write_dataset()` function from the arrow package, which respects the grouping created by the dplyr `group_by()` function.
+To create a multi-file data set with one parquet file per year of library data we would do the following:
+
+```{r write-dataset}
+#| eval: !expr "!file.exists('data/seattle-library-checkouts/CheckoutYear=2021/part-0.parquet')"
+seattle_lib |>
+  group_by(CheckoutYear) |>
+  write_dataset(
+    path = "data/seattle-library-checkouts",
+    format = "parquet"
+  )
+```
+
+Let's take a look at the files:
+
+```{r show-parquet-files}
+seattle_lib_files <- list.files(
+  path = "data/seattle-library-checkouts",
+  recursive = TRUE
+) 
+seattle_lib_files
+```
+
+These filenames follow a "self-describing" convention used by the Apache Hive project ([hive.apache.org](https://hive.apache.org/)).
+Hive-style partitions name folders using a "key=value" convention, and the data files in that folder contain the subset of the data for which the key has the relevant value.
+Following this convention makes it easier to locate files relevant to a particular query.
+
+### Performance advantages
+
+Partitioning a large data set into many smaller, meaningful files can produce substantial improvements in performance.
+To illustrate, we'll repeat the analysis of the 2021 Seattle library book checkouts by month using the partitioned data:
+
+```{r dataset-performance-multiple-parquet}
+#| cache: true
+system.time({
+  open_dataset(
+    sources = "data/seattle-library-checkouts/", 
+    format = "parquet"
+  ) |>
+    filter(CheckoutYear == 2021, MaterialType == "BOOK") |>
+    group_by(CheckoutMonth) |>
+    summarise(TotalCheckouts = sum(Checkouts)) |>
+    arrange(desc(CheckoutMonth)) |>
+    collect()
+})
+```
+
+This is a 10-fold speed up compared to a single parquet file, and a 100-fold speed up compared to the original csv file.
+
+## Other topics
+
+I haven't quite worked out the best way to integrate these sections.
+
+### Using duckdb with arrow
+
+One useful feature of arrow is that it is interoperable with duckdb and dbplyr.
```

Comment:
Maybe this could go be a subsection in "Using dplyr with arrow"? With some connection like:

> There's also an even more direct connection between arrow and dbplyr — it's very easy to turn an arrow dataset into a duckdb datasource. This doesn't involve any memory copying, and speaks to the goals of the arrow ecosystem: enable seamless transitions from one computing framework to another.